docs/tech-specs/basic-spec.md

```diff
@@ -6,6 +6,269 @@
+
+## Implementation Details
+
+### Objects
+
+![Class Diagram](./images/basic-spec/BrokerObjects.svg)
+
+### Endpoints
+
+We want the broker service itself to be stateless and follow the RESTful API paradigms. As follows are the conventions for the API -
+
+* `PUT` for creating **Producer**, **Channel** and **Consumer**; primarily to allow the client to dictate the ID; it will be idempotent so it can work as edit as well
+* All `PUT` request will accept `application/x-www-form-urlencoded` MIME Type Form data
+* All the `GET` will expose a `Last-Modified` header
+* `DELETE` is not supported for any resource type
+* All list endpoints will be ordered by ID and hence use ID as the pagination key
+* When `PUT` requests will created it will return `201`; else `200` when updated.
+* **Message** can only be a `POST` and can not be amended, but can be fetched.
+  * The POST call will need to come with special headers to identify Producer
+    * `X-Broker-Producer-ID` ID of the Producer
+    * `X-Broker-Producer-Token` Token of the Producer
+  * The POST call also will need the channel token for simple auth purpose - `X-Broker-Channel-Token`
+  * The message priority will also need to be passed via header - `X-Broker-Message-Priority`
+  * Message's content type will be derived from request `Content-Type`; if missing will default to `application/octet-stream`
+* The **Message** list endpoint can be filtered using _statusChangedSince_ query parameter
+* The **Message** `GET` endpoint will list all the jobs and their status in the resource itself since the **Message** and **DeliverJob** are both immutable through the API.
+  * **DeliveryJob** will expose re-trigger endpoint to be used by consumer, but only if the delivery status is _dead_.
+* **Message** delivery or **DeliveryJob** can be re-triggered using _Admin Token_ by the Fail-safe worker
+  * That means they will have special re-trigger endpoint which will need _Admin Token_.
+  * DLQ'd jobs can be re-triggered by consumer using its _Consumer Token_.
+  * The Admin token is - `X-Broker-Admin-Token`
+  * The delivery to a consumer will also contain `X-Broker-Consumer-Token` to ensure the request is coming from Broker, in addition to a custom `User-Agent`
+
+So the endpoints available would be -
+
+1. PUT /producer/{producer-id}
```

Comment:
Why not follow REST if this API is so similar? Otherwise clients need to generate IDs when creating resources

Reply:
The idea here is, client is responsible for the unique producer id, but the token is something the system generates; so in such situation producer will need to send the combination of ID and Token to identify itself during broadcast. When it comes to making it configuration based creation - both ID and Token can be provided by the system configurator.

Reply:
@ahmedshuhel it does but there was no `POST /producer` API, so in order to do the same you have to come up with your own producer ID and use PUT